src/script/lua_api/l_env.cpp

```diff
@@ -880,6 +880,21 @@ int ModApiEnvMod::l_find_node_near(lua_State *L)
 	return 0;
 }
 
+static void checkArea(v3s16 &minp, v3s16 &maxp)
+{
+	auto volume = VoxelArea(minp, maxp).getVolume();
+	// Volume limit equal to 8 default mapchunks, (80 * 2) ^ 3 = 4,096,000
+	if (volume > 4096000) {
+		throw LuaError("Area volume exceeds allowed value of 4096000");
+	}
+
+	// Clamp to map range to avoid problems
+#define CLAMP(arg) core::clamp(arg, (s16)-MAX_MAP_GENERATION_LIMIT, (s16)MAX_MAP_GENERATION_LIMIT)
```

Comment:
I have figured out that for Lua selftest purposes related to the Lua workaround for the crash this PR is addressing I require an area which is slightly out of MAX_MAP_GENERATION_LIMIT, but can hold nodes. This area exists in current Minetest and this patch seems to introduce an area in which unit tests misbehave.

I once again want to urge you to not re-use MAX_MAP_GENERATION_LIMIT here – it has nothing to do with the crash and is a solution that is simple, obvious, and wrong. Use the boundaries of the underlying s16 data type please to avoid both creating and masking additional bugs.

**Sigh.**

Reply:
There's no point in supporting map areas outside this area. `MAX_MAP_GENERATION_LIMIT` was exactly created to avoid such over- and underflow issues in the map generation. Hence it makes perfect sense to also apply those rules to the rest of the API. You can still limit the map generation further using the map_meta.txt settings.

The overflow is likely to happen in 
```C++
		for (p.X = minp.X; p.X <= maxp.X; p.X++)
		for (p.Y = minp.Y; p.Y <= maxp.Y; p.Y++)
		for (p.Z = minp.Z; p.Z <= maxp.Z; p.Z++) {
```

where maxp equals S16_MAX. There must be no area outside of `MAX_MAP_GENERATION_LIMIT`, hence this is a sane solution.

Reply:
> There's no point in supporting map areas outside this area.

In abstract: By clamping at anything else than the boundaries of the underlying data type you guarantee that Minetest has a function that behaves well only for a subset of the values representable by the data type.

One problem resulting from such a construction (boundaries do not align with underlying data type) exists already – the engine can return positions for players that make it fall over and die when those are given to a Lua function that takes a position: https://github.com/minetest/minetest/issues/11742?

I am trying to prevent more of these kind of mistakes that are essentially type errors.

> MAX_MAP_GENERATION_LIMIT was exactly created to avoid such over- and underflow issues in the map generation.

@sfan5 showed me a comment about that already. I have verified that stopping at MAX_MAP_GENERATION_LIMIT avoids encountering a bunch of overflows during map generation. The problem here is that these overflows still exist and can be abused by attackers if the functions can be called from outside mapgen.

**Masking, i.e. not triggering bugs in mapgen does not magically make them vanish.**

> There must be no area outside of MAX_MAP_GENERATION_LIMIT, hence this is a sane solution.

Ex falso quodlibet – tHe above statement is about as true as the following:

If the Eiffel tower is located in New York, the moon is made out of cheese.

Reply:
> Another problem resulting from such a construction (boundaries do not align with underlying data type) has been found already: #11742

No. The problem highlighted by the other issue exists *because* there is no clamping for the player positions (you also said you don't like that solution).

> The problem here is that these overflows still exist and can be abused by attackers if the functions can be called from outside mapgen.

What does that mean?
The behaviour of `find_nodes_in_area` is exactly the same with this PR inside and outside of mapgen so any claimed difference that could be abused does not exist.

> Ex [...]

Is that a magic chant that allows you to discard an argument without losing a single word on its contents?
Gotta remember that one.

Reply:
> Is that a magic chant that allows you to discard an argument without losing a single word on its contents?
Gotta remember that one.

It's literally just saying that a logical implication based on a false premise allows you to conclude anything. In other words, erlehmann is saying that

> There must be no area outside of MAX_MAP_GENERATION_LIMIT

is false, as clearly placing nodes there is currently possible for malicious actors; and therefore the conclusion

> hence this is a sane solution.

can't be drawn.

Reply:
> unnecessarily quirky

Sorry for that, I think I am just overly annoyed that ppl repeat that stuff.

> clearly placing nodes there is currently possible for malicious actors

Worse though – even if it was not possible, they do not even need to.

(It is possible though and I know another crash that I can trigger.)

Reply:
Perhaps the proper solution is to fix this, making it impossible for players to interact with area outside of `MAX_MAP_GENERATION_LIMIT`?

Reply:
> It's literally just saying that a logical implication based on a false premise allows you to conclude anything

Yes, I know what it means and that it's an axiom. But I didn't see anything about *why* the premise is supposed to be incorrect.
Although nodes could theoretically exist up to (+/-)31007, in my testing the map always ends far before 31000.